### PR TITLE
Added code for Version Check

### PR DESCRIPTION
**Pull request by Priya411**
Code review for Version Check activity

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -181,6 +185,23 @@ private UserContext findUser(String name) {
...
     });
 
+    // Added by Priyanka Agarwal according to CodeU Tutorial
+    // Added import statement
+    panel.register("info", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        final ServerInfo info = context.getInfo();
+        if (info == null) {
+          // Communicate error to user - the server did not send us a valid
```
Comment (julezjw) [inline src/codeu/chat/client/commandline/Chat.java:201]: Hmm, I think you're supposed to add in code to execute what currently commented out

Comment (Priya411) [reply src/codeu/chat/client/commandline/Chat.java:201]: You're correct, I'll change that.

**Comment by KevinWorkman** [comment]
Just a couple of general notes: I'd try to have more descriptive pull request titles, like "Added version check feature" or something.

Also, in the description of this pull request I'd add instructions for actually using the feature. How do I check to make sure this is working? How do I issue a version check command?

**Comment by Priya411** [comment]
The way we can access the Version is by running the Client. If you type help, there is now an added option of Info, and when you type Info, you can now see the version number.

File: src/codeu/chat/client/core/Context.java [hunk 2]
```diff
@@ -22,15 +22,19 @@
 import codeu.chat.common.User;
 import codeu.chat.util.Uuid;
 import codeu.chat.util.connections.ConnectionSource;
+import codeu.chat.common.ServerInfo;
+import codeu.chat.client.core.View;
 
 public final class Context {
 
   private final BasicView view;
+  private final View view1;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/core/Context.java]: I would use a more descriptive variable name here. Why do we need both the `view` and the `view1` variables? Aren't they basically the same thing?

File: src/codeu/chat/client/core/Context.java [hunk 3]
```diff
@@ -40,6 +44,12 @@ public UserContext create(String name) {
         new UserContext(user, view, controller);
   }
 
+  // Added according to CodeU tutorial by Priyanka Agarwal
+  // Created View1 Variable
```
Comment (KevinWorkman) [inline src/codeu/chat/client/core/Context.java]: I would have a comment explaining what this function returns, not that it was added from a tutorial. The tutorial part is less important to somebody reading the code than a comment explaining what the function returns.

File: src/codeu/chat/client/core/View.java [hunk 4]
```diff
@@ -136,4 +137,23 @@ public View(ConnectionSource source) {
...
+
+  // Added by Priyanka Agarwal according to CodeU tutorial
+  public ServerInfo getInfo() {
+    try (final Connection connection = this.source.connect()) {
+      Serializers.INTEGER.write(connection.out(), NetworkCode.SERVER_INFO_REQUEST);
+      if (Serializers.INTEGER.read(connection.in()) == NetworkCode.SERVER_INFO_RESPONSE) {
+        final Uuid version = Uuid.SERIALIZER.read(connection.in());
+        return new ServerInfo(version);
+      } else {
+        // Communicate this error - the server did not respond with the type of
```
Comment (KevinWorkman) [inline src/codeu/chat/client/core/View.java:152]: Does something special need to happen here? Or is it handled by `return null` below? If it's handled below, maybe modify this comment to explain that?

File: src/codeu/chat/common/ServerInfo.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*Class created by Priyanka Agarwal as per code written
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfo.java]: You don't have to comment saying this is from the tutorial. I'd maybe add comments explaining what this class is for though.

File: src/codeu/chat/common/ServerInfo.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+on codeU directions. Only added package and import statements
+ */
+
+package codeu.chat.common;
+import codeu.chat.util.Uuid;
+import java.io.IOException;
+
+
+public final class ServerInfo {
+    private final static String SERVER_VERSION = "1.0.0";
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfo.java]: Maybe a comment or documentation somewhere saying that this variable should be updated whenever the server code changes?

File: src/codeu/chat/client/commandline/Chat.java [hunk 7]
```diff
@@ -97,6 +98,7 @@ private Panel createRootPanel(final Context context) {
     // Add a command to print a list of all commands and their description when
     // the user for "help" while on the root panel.
     //
+    // Added info command information: Priyanka Agarwal
```
Comment (KevinWorkman) [inline src/codeu/chat/client/commandline/Chat.java]: This comment probably belongs below, when you actually add the info command?

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (KevinWorkman) [reply src/codeu/chat/client/commandline/Chat.java:201]: Minor note: try to comment *after* you've changed something, that way the commenter knows when to come back and look at your code. It's hard to know when to come back if you say you *will* change it. It's a lot easier if you say "I fixed that, please take another look" instead.

File: src/codeu/chat/common/ServerInfo.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        } catch (IOException e){
+            System.out.println("Invalid input");
+        }
+    }
+    public ServerInfo(Uuid version) {
+        this.version = version;
+    }
+
+    public String toString()
+    {
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfo.java]: Minor note: The above functions use a curly bracket on the same line as the function, but this one puts it on another line. I'd say try to stick with one style.

File: src/codeu/chat/common/ServerInfo.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+package codeu.chat.common;
+import codeu.chat.util.Uuid;
+import java.io.IOException;
+
+
+public final class ServerInfo {
+    private final static String SERVER_VERSION = "1.0.0";
+
+    public Uuid version;
+    public ServerInfo() {
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfo.java]: You might think about adding a test class for this class? Even if it's not in the tutorial, it's a good habit to get into. Let me know if you have questions about that. Or I think Matt's activity includes adding a test, so maybe talk to him too?

**Review by KevinWorkman — COMMENTED** [review]
Looks pretty good. I've added a couple comments below.

I tend to phrase my comments as a question. That doesn't necessarily mean I need a reply with an answer- instead, maybe think about how the code might be structured to answer the question itself.

File: src/codeu/chat/common/ServerInfo.java (hunk 6 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfo.java]: This has been fixed, please let me know if I can improve this more.

File: src/codeu/chat/client/commandline/Chat.java (hunk 7 above)
Comment (Priya411) [reply src/codeu/chat/client/commandline/Chat.java]: I changed the positioning of the comments and added more comments to give more information about the code I added.

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (Priya411) [reply src/codeu/chat/client/commandline/Chat.java:201]: Thanks for the feedback, I'll make sure to write the rest of my comments like that.

File: src/codeu/chat/common/ServerInfo.java (hunk 5 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfo.java]: I added some more comments on the top of the class as well as for the functions to add clarity. Please let me know if I can improve this.

File: src/codeu/chat/common/ServerInfo.java (hunk 8 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfo.java]: I made the style the same with the bracket on the same line as the function.

File: src/codeu/chat/client/core/Context.java (hunk 2 above)
Comment (Priya411) [reply src/codeu/chat/client/core/Context.java]: I changed the variable name and added an explanation for why I added the second view variable

File: src/codeu/chat/client/core/Context.java (hunk 3 above)
Comment (Priya411) [reply src/codeu/chat/client/core/Context.java]: I added a comment regarding the purpose of this function

File: src/codeu/chat/client/core/View.java (hunk 4 above)
Comment (Priya411) [reply src/codeu/chat/client/core/View.java:152]: I added comments to explain the function and print lines to display the error information. Please let me know if I can improve this.

File: src/codeu/chat/common/ServerInfo.java (hunk 9 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfo.java]: I added a test class to ensure the versions were correct. Please let me know if I can add more to it

File: src/codeu/chat/common/ServerInfoTest.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import org.junit.Test;
+import codeu.chat.util.Uuid;
+
+import java.io.IOException;
+
+public class ServerInfoTest {
+
+    @Test
+    public void testServerInfo() {
+        ServerInfo server1 = new ServerInfo();
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfoTest.java]: It's awesome to see a test class! Nice work!

The only thing I'm worried about is now anytime anybody changes the server version, they'll also have to change the test class. Maybe the `ServerInfo` class should always take the version as a parameter? What do you think?

File: src/codeu/chat/client/core/Context.java [hunk 11]
```diff
@@ -22,15 +22,23 @@
...
+  // the method that returns this is private in the child class 
+  // BasicView, we're going to create a variable ViewInterface,
+  // which has a public version of this function - Priyanka Agarwal
+  private final View viewInterface;
   private final Controller controller;
 
   public Context(ConnectionSource source) {
     this.view = new View(source);
     this.controller = new Controller(source);
+    this.viewInterface = new View(source);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/core/Context.java]: I'm still a little bit confused about why you need two instances of `View` here. I think I understand why you did it: because the `getInfo()` function is in the `View` class, not the `BasicView` interface, so you needed a reference to a `View` variable?

But note that the pre-existing `view` variable is already a `View` instance, so it does contain the `getInfo()` class. You just need to reference it using the subclass. The easiest ways to do that would be to just change the `view` variable to a `View` reference (change line 30 to `private final View view;`), or you could just cast it before using it (that way is a little bit messier though).

Another option would be to just add the `getInfo()` function to the `BasicView` interface. I'm not sure how feasible that is though- depends on how many other classes implement the `BasicView` interface.

Does that make sense?

Comment (Priya411) [reply src/codeu/chat/client/core/Context.java]: I removed the viewInterface object and have made everything into a object called view of type View

File: src/codeu/chat/common/ServerInfoTest.java (hunk 10 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfoTest.java]: I'm a little bit confused as to what you mean by this. For the parameter, is this for the constructor?

Comment (Priya411) [reply src/codeu/chat/common/ServerInfoTest.java]: The way I understood this, was the 1.0.0 was the default version if there's no parameters. If the server version is changed, which is when there are now parameters, the version will be what the parameters are. Should I leave the empty parameter test case as is, but then just use a variable for the version and check accordingly so that the version can be changed for the test case with the parameter?

File: src/codeu/chat/server/Server.java [hunk 12]
```diff
@@ -53,10 +54,9 @@
   private final Timeline timeline = new Timeline();
 
   private final Map<Integer, Command> commands = new HashMap<>();
-
   private final Uuid id;
   private final Secret secret;
-
+  private static final ServerInfo info = new ServerInfo();
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Server.java]: This is how the `ServerInfo` is created **on the server side.** You probably want to pass in a parameter here?

Basically, you're going to want to update the server version variable whenever you make a change to the server code. Where you do that is up to you: either by passing a parameter in, or by changing the variable inside the `ServerInfo` class. But you want to do that in a way that's easy to test, too.

So right now you either want to add a parameter here, or you want to change your test class so it doesn't break when people change the variable.

Does that make more sense?

File: src/codeu/chat/client/core/View.java [hunk 13]
```diff
@@ -136,4 +137,28 @@ public View(ConnectionSource source) {
...
+  // Added by Priyanka Agarwal
+  // This function returns the information about the Server,
+  // including it's Versions based on the specific server it
+  // is using and is connected to. 
+  public ServerInfo getInfo() {
+    try (final Connection connection = this.source.connect()) {
+      Serializers.INTEGER.write(connection.out(), NetworkCode.SERVER_INFO_REQUEST);
+      if (Serializers.INTEGER.read(connection.in()) == NetworkCode.SERVER_INFO_RESPONSE) {
+        final Uuid version = Uuid.SERIALIZER.read(connection.in());
+        return new ServerInfo(version);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/core/View.java:150]: Note: This is creating a `ServerInfo` **on the client side** based on values it's getting from the server side. So the other half of this coin is: how is the `ServerInfo` being created on the server side? (See my comment in the `Server` class in this review.)

File: src/codeu/chat/common/ServerInfoTest.java (hunk 10 above)
Comment (KevinWorkman) [reply src/codeu/chat/common/ServerInfoTest.java]: I see what you're saying.

Hopefully my other comments in this review are a little more clear, but basically there are two different places a `ServerInfo` is created:

- First it's created on the server side inside the `Server` class.
- Then the value inside that instance is sent to the client side, and the client creates its own instance.

The point of this is, whenever you make a change to the server code, you want to change the value (either the parameter or the default variable) of that first instance. That gets sent to the client, so the client can know what version the server is on. (This might become useful later as you start adding features.)

And I'm wondering about this test because I'm not sure how you're going to be changing the value on the server side. Right now the server side always creates its instance using the default no-args constructor.

Does that make more sense?

File: src/codeu/chat/common/ServerInfo.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+displayed in the Panels, to be use by the Chat.java class.
+ */
+
+package codeu.chat.common;
+import codeu.chat.util.Uuid;
+
+
+public final class ServerInfo {
+
+    public Uuid version;
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfo.java]: Does this need to be public? Maybe add JavaDoc?

File: src/codeu/chat/common/ServerInfoTest.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+public class ServerInfoTest {
+
+    @Test
+    public void testServerInfo() {
+        try{
+            ServerInfo server2 = new ServerInfo(Uuid.parse("1.0.1"));
+            assertEquals("1.0.1", server2.toString());
+        }
+        catch (IOException e){
+            System.out.println("Invalid input");
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfoTest.java]: I would make sure that this test fails if this exception is hit.

File: src/codeu/chat/server/Server.java [hunk 16]
```diff
@@ -53,10 +54,17 @@
...
 
   private final Map<Integer, Command> commands = new HashMap<>();
-
   private final Uuid id;
   private final Secret secret;
+  private ServerInfo info = null;
 
+  {
+    try {
+      info = new ServerInfo(Uuid.parse("1.0.0"));
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Server.java]: I would add a note making sure people know to change this variable whenever they change the server code. Or maybe move the `String` into its own variable?

Actually now that I'm looking at it, I realize it would probably be better to just move this whole thing inside the `Server` constructor instead of using an initialization block. I'll leave the final decision up to you though.

**Review by KevinWorkman — APPROVED** [review]
LGTM (looks good to me)

I've left a few minor notes, but I'll leave the final decisions up to you.

One final thing: make sure this works with Matt and Julia's changes as you start merging them. Specifically, Matt's changes modify how you issues commands, and Julia's changes also added messages.

File: src/codeu/chat/server/Server.java (hunk 16 above)
Comment (Priya411) [reply src/codeu/chat/server/Server.java]: I moved the initialization to the constructor and created a String variable for the version, which the user can changed based on what they want the server version to be. I also added comments explaining these two and how they can be customized.

File: src/codeu/chat/common/ServerInfo.java (hunk 14 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfo.java]: I added JavaDoc under the comments on the top.

File: src/codeu/chat/common/ServerInfoTest.java (hunk 15 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfoTest.java]: I added three more tests, two which should return an Exception, and one with a negative version

File: src/codeu/chat/server/Server.java [hunk 17]
```diff
@@ -53,23 +54,31 @@
...
+  private ServerInfo info = null;
   private final Model model = new Model();
   private final View view = new View(model);
   private final Controller controller;
-
   private final Relay relay;
   private Uuid lastSeen = Uuid.NULL;
+  // This string should be changed whenever the initial version value for the server
+  // needs to be changed
+  private String version = "1.0.0";
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Server.java]: small note: this is no longer the initial version of the server, is it? This is the current version of the server.

File: src/codeu/chat/server/Server.java [hunk 18]
```diff
@@ -53,23 +54,31 @@
...
     this.id = id;
     this.secret = secret;
     this.controller = new Controller(id, model);
     this.relay = relay;
+    // This try catch block initializes the server version based on the version
+    // set in object 'version'
+    try {
+      info = new ServerInfo(Uuid.parse(version));
+    } catch (IOException e) {
+      System.out.println("Invalid input");
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Server.java:81]: This exception shouldn't ever happen in real life, right? I'd add info the the `println()` saying to check the version number, since that's the only way to cause this exception.

File: src/codeu/chat/common/ServerInfoTest.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        try {
+            // Exception only caught when version is larger than 32 bits
+            ServerInfo server3 = new ServerInfo(Uuid.parse("1.999999999999999999"));
+            fail(); // if we got here, no exception was thrown, which is bad
+        }
+        catch (IOException e) {
+            assertEquals(true,true);
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ServerInfoTest.java]: I think there's a better way to make sure an exception is thrown. Check out: https://stackoverflow.com/questions/156503/how-do-you-assert-that-a-certain-exception-is-thrown-in-junit-4-tests

File: src/codeu/chat/common/ServerInfo.java (hunk 14 above)
Comment (KevinWorkman) [reply src/codeu/chat/common/ServerInfo.java]: Hmm, I'm still not sure why it needs to be public? And minor note with the comments at the top: You no longer have a default version, right? Since you always use the parameterized constructor?

**Review by KevinWorkman — APPROVED** [review]
I've added a few minor comments, but this still LGTM. I'd say coordinate with your teammates to start merging this in whenever you're ready.

File: src/codeu/chat/common/ServerInfoTest.java (hunk 19 above)
Comment (Priya411) [reply src/codeu/chat/common/ServerInfoTest.java]: I used your suggested way to do the testing for the exception. Thanks for the link! The examples I found online were really confusing, so this helped a lot.

File: src/codeu/chat/server/Server.java (hunk 17 above)
Comment (Priya411) [reply src/codeu/chat/server/Server.java]: Yes sorry, you're correct. I updated the comments to be accurate.

File: src/codeu/chat/server/Server.java (hunk 18 above)
Comment (Priya411) [reply src/codeu/chat/server/Server.java:81]: I added a println to give the user more information about what would cause this error

File: src/codeu/chat/common/ServerInfoTest.java (hunk 19 above)
Comment (KevinWorkman) [reply src/codeu/chat/common/ServerInfoTest.java]: No problem! It might be a little easier if you split each test into its own function.
